# nsoindia - Examples

Python client for India's National Statistical Office (MoSPI) data portal.
Access indicators across 19 datasets: employment, prices, GDP, health, education, environment, and more.

```bash
pip install nsoindia[pandas]
```

In [1]:
import warnings
warnings.filterwarnings('ignore')
import nsoindia
from nsoindia.exceptions import InvalidDatasetError, InvalidFilterError, APIError, NoDataError

---
## Available Datasets

In [31]:
nsoindia.list_datasets(format="df")

""


---
## PLFS - Periodic Labour Force Survey

Covers labour force participation, unemployment, worker population ratio, wages.
Use `get_indicators` to see what indicator codes are available, `get_metadata` to see valid filter values for a given indicator.

In [3]:
# Indicators grouped by frequency_code
indicators = nsoindia.get_indicators("PLFS")
for group, items in indicators["indicators_by_frequency"].items():
    print(f"\n{group}:")
    for item in items:
        print(f"  {item['indicator_code']}: {item.get('indicator_name', item.get('description', ''))}")


frequency_code_1_Annual:
  1: LFPR (Labour Force Participation Rate, in per cent)
  2: WPR (Worker Population Ratio, in per cent)
  3: UR (Unemployment Rate, in per cent)
  4: Percentage distribution of workers
  5: Percentage of regular wage/ salaried employees with employment condition as
  6: Average wage/salary earnings (Rs. 0.00) during the preceding calendar month from regular wage/salaried employment
  7: Average wage earnings (Rs. 0.00) per day from casual labour work other than public works
  8: Average gross earnings (Rs. 0.00) during last 30 days from self-employment

frequency_code_2_Quarterly:
  1: LFPR (Labour Force Participation Rate, in per cent)
  2: WPR (Worker Population Ratio, in per cent)
  3: UR (Unemployment Rate, in per cent)
  4: Percentage distribution of workers

frequency_code_3_Monthly:
  1: LFPR (Labour Force Participation Rate, in per cent)
  2: WPR (Worker Population Ratio, in per cent)
  3: UR (Unemployment Rate, in per cent)


In [4]:
# Valid filter values for Unemployment Rate (indicator_code=3), Annual (frequency_code=1)
metadata = nsoindia.get_metadata("PLFS", indicator_code=3, frequency_code=1,)
for key, values in metadata["filter_values"]["data"].items():
    print(f"\n{key}:")
    for v in values[:4]:
        print(f"  {v}")


year:
  {'year': '2017-18'}
  {'year': '2018-19'}
  {'year': '2019-20'}
  {'year': '2020-21'}

age:
  {'age_code': 1, 'description': '15 years and above'}
  {'age_code': 2, 'description': '15-29 years'}
  {'age_code': 3, 'description': '15-59 years'}
  {'age_code': 4, 'description': 'all'}

education:
  {'description': '1.not literate', 'education_code': 1}
  {'description': '2.literate & upto primary', 'education_code': 2}
  {'description': '3.middle', 'education_code': 3}
  {'description': '4.secondary', 'education_code': 4}

gender:
  {'description': 'female', 'gender_code': 2}
  {'description': 'male', 'gender_code': 1}
  {'description': 'person', 'gender_code': 3}

religion:
  {'description': 'all', 'religion_code': 1}
  {'description': 'Christianity', 'religion_code': 4}
  {'description': 'Hinduism', 'religion_code': 2}
  {'description': 'Islam', 'religion_code': 3}

sector:
  {'description': 'rural', 'sector_code': 1}
  {'description': 'rural + urban', 'sector_code': 3}
  {'des

In [5]:
# Unemployment Rate - All India, all years, both sectors
nsoindia.get_data("PLFS", {
    "indicator_code": 3,
    "frequency_code": 1,
    "state_code": 99,     # 99 = All India
    "gender_code": 3,     # 3 = Person
    "age_code": 1,        # 1 = 15 years and above
    "sector_code": 3,     # 3 = rural + urban
}, format="df")

,year,frequency,indicator,state,gender,sector,AgeGroup,weekly_status,religion,socialGroup,General_Education,quarter,month,value,unit
0,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,Christianity,all,all,all,None,8.4,%
1,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,Hinduism,all,all,all,None,5.8,%
2,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,other backward class,all,all,None,5.9,%
3,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,Islam,all,all,all,None,7.2,%
4,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,all,all,None,6.0,%
5,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,others,all,all,None,6.7,%
6,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,CWS,all,all,all,all,None,8.7,%
7,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,scheduled tribe,all,all,None,4.2,%
8,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,scheduled caste,all,all,None,6.2,%
9,2017-18,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,Sikhism,all,all,all,None,6.8,%


In [21]:
# Average wages - All India, annual, 2023-24
nsoindia.get_data("PLFS", {
    "indicator_code": 2,   # WPR (Worker Population Ratio, in per cent)
    "frequency_code": 1,
    "year": "2023-24",
    "state_code": 99,
    "gender_code": 3,
    "age_code": 1,
    "sector_code": 3,
}, format="df")

,year,frequency,indicator,state,gender,sector,AgeGroup,weekly_status,religion,socialGroup,General_Education,quarter,month,value,unit
0,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,all,all,None,58.2,%
1,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,CWS,all,all,all,all,None,53.7,%
2,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,1.not literate,all,None,59.6,%
3,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,2.literate & upto primary,all,None,68.5,%
4,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,3.middle,all,None,60.7,%
5,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,4.secondary,all,None,49.7,%
6,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,5.higher secondary,all,None,45.9,%
7,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,6.diploma/ certificate course,all,None,73.6,%
8,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,7.graduate,all,None,57.5,%
9,2023-24,Annually,"WPR (Worker Population Ratio, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,8.post graduate & above,all,None,62.2,%


---
## CPI - Consumer Price Index

Multiple base years (2010, 2012, 2024) and two levels: Group and Item.

In [23]:
# Available base years, levels, and series
nsoindia.get_indicators("CPI", format="df")

,base_year,level,series
0,2010,NaN,NaN
1,2012,NaN,NaN
2,2024,NaN,NaN
3,NaN,Group,NaN
4,NaN,Item,NaN
5,NaN,NaN,Current
6,NaN,NaN,Back


In [25]:
# Filter values for base_year=2012, Group level
metadata = nsoindia.get_metadata("CPI", base_year="2012", level="Group", series="Current", format="df")
metadata

,series,year,month_code,month_name,state_code,state_name,sector_code,sector_name,group_code,group_name,subgroup_code,subgroup_name
0,Current,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Back,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,2025.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,NaN,2024.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,2023.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
100,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6,NaN,6.1.04,Recreation and Amusement
101,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6,NaN,6.1.05,Education
102,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6,NaN,6.1.06,Personal Care and Effects
103,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,6,NaN,6.99,Miscellaneous-Overall


In [26]:
# CPI Group index, base year 2012, year 2024
nsoindia.get_data("CPI", {
    "base_year": "2012",
    "series": "Current",
    "year": "2024",
}, format="df")

,baseyear,year,month,state,sector,group,subgroup,index,inflation,status
0,2012,2024,December,All India,Rural,Consumer Food Price,Consumer Food Price-Overall,204.7,8.65,F
1,2012,2024,December,All India,Rural,Miscellaneous,Transport and Communication,176.7,2.85,F
2,2012,2024,December,All India,Rural,Miscellaneous,Education,192.2,3.72,F
3,2012,2024,December,All India,Rural,Miscellaneous,Miscellaneous-Overall,190.8,4.26,F
4,2012,2024,December,All India,Rural,Miscellaneous,Household Goods and Services,187.0,2.47,F
5,2012,2024,December,All India,Rural,Miscellaneous,Health,200.2,4.00,F
6,2012,2024,December,All India,Rural,Miscellaneous,Personal Care and Effects,206.3,9.68,F
7,2012,2024,December,All India,Rural,Miscellaneous,Recreation and Amusement,181.5,2.54,F
8,2012,2024,December,All India,Rural,Fuel and Light,Fuel and Light-Overall,182.3,-0.44,F
9,2012,2024,December,All India,Rural,Housing,Housing-Overall,None,None,F


---
## NAS - National Accounts Statistics (GDP)

Annual and quarterly GDP, GVA, consumption, capital formation by industry and institutional sector.

In [27]:
nsoindia.get_indicators("NAS", format="df")

,indicator_code,description
0,1,Gross Value Added
1,2,Net Taxes on Products
2,3,Taxes on Products
3,4,Subsidies on Products
4,5,Gross Domestic Product
5,6,Consumption of Fixed Capital
6,7,Net Domestic Product
7,8,Gross Capital Formation by Industry of Use
8,9,Gross Fixed Capital Formation
9,10,Private Final Consumption Expenditure


In [28]:
# Filter values for GVA by industry (indicator_code=1), base year 2022-23
metadata = nsoindia.get_metadata("NAS", indicator_code=1, base_year="2022-23", series="Current", frequency_code=1)
for key, values in metadata.items():
    if isinstance(values, list):
        print(f"{key}: {values[:3]}")

year: [{'year': '2025-26'}, {'year': '2024-25'}, {'year': '2023-24'}]
revision: [{'revision_code': 2, 'revision_name': 'Second Advance Estimates'}, {'revision_code': 4, 'revision_name': 'First Revised Estimates'}, {'revision_code': 8, 'revision_name': 'Final Estimates'}]
industry: [{'industry_code': 1, 'industry_name': 'Agriculture, Livestock, Forestry and Fishing'}, {'industry_code': 2, 'industry_name': 'Mining and Quarrying'}, {'industry_code': 3, 'industry_name': 'Manufacturing'}]
subindustry: [{'subindustry_code': 1, 'subindustry_name': 'Crops', 'industry_code': 1}, {'subindustry_code': 2, 'subindustry_name': 'Livestock', 'industry_code': 1}, {'subindustry_code': 3, 'subindustry_name': 'Forestry and Logging', 'industry_code': 1}]
institutional_sector: [{'institutional_code': 1, 'institutional_name': 'Public Non-Financial Corporations'}, {'institutional_code': 2, 'institutional_name': 'Private Non-Financial Corporations'}, {'institutional_code': 3, 'institutional_name': 'Public Fina

In [29]:
# GVA by industry, current prices, latest estimates
nsoindia.get_data("NAS", {
    "indicator_code": 1,
    "base_year": "2022-23",
    "series": "Current",
    "frequency_code": 1,
    "year": "2024-25",
}, format="df")

,base_year,series,year,indicator,frequency,revision,industry,subindustry,institutional_sector,quarter,current_price,constant_price,unit
0,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Public Non-Financial Corporations,None,1551904.3111790416,None,₹ Crore
1,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Private Non-Financial Corporations,None,9555601.402799515,None,₹ Crore
2,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Public Financial Corporations,None,780342.4093269374,None,₹ Crore
3,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Private Financial Corporations,None,1059339.3228545364,None,₹ Crore
4,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,General Government,None,2656799.960246522,None,₹ Crore
5,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Households Including NPISH,None,13250480.021178257,None,₹ Crore
6,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,None,None,Total Gross Value Added,None,28854467.42758481,None,₹ Crore
7,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,"Agriculture, Livestock, Forestry and Fishing",Crops,None,None,3487514.5392785612,3113280.9316670736,₹ Crore
8,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,"Agriculture, Livestock, Forestry and Fishing",Livestock,None,None,1387062.690777359,1278634.3400525874,₹ Crore
9,2022-23,Current,2024-25,Gross Value Added,Annual,First Revised Estimates,"Agriculture, Livestock, Forestry and Fishing",Forestry and Logging,None,None,408682.0267620707,414487.83170621865,₹ Crore


---
## IIP - Index of Industrial Production

Monthly and annual industrial output across manufacturing, mining, and electricity sectors.

In [30]:
# Available categories and years for IIP annual
metadata = nsoindia.get_metadata("IIP", base_year="2011-12", frequency="Annually")
metadata

{'financial_year': [{'financial_year': '2024-25'},
  {'financial_year': '2023-24'},
  {'financial_year': '2022-23'},
  {'financial_year': '2021-22'},
  {'financial_year': '2020-21'},
  {'financial_year': '2019-20'},
  {'financial_year': '2018-19'},
  {'financial_year': '2017-18'},
  {'financial_year': '2016-17'},
  {'financial_year': '2015-16'},
  {'financial_year': '2014-15'},
  {'financial_year': '2013-14'},
  {'financial_year': '2012-13'}],
 'type': [{'type': 'Use-based category'},
  {'type': 'Sectoral'},
  {'type': 'General'}],
 'category': [{'category_code': 1,
   'category_name': 'Mining',
   'type': 'Sectoral'},
  {'category_code': 2, 'category_name': 'Manufacturing', 'type': 'Sectoral'},
  {'category_code': 3, 'category_name': 'Electricity', 'type': 'Sectoral'},
  {'category_code': 4, 'category_name': 'General', 'type': 'General'},
  {'category_code': 5,
   'category_name': 'Primary Goods',
   'type': 'Use-based category'},
  {'category_code': 6,
   'category_name': 'Capital Go

In [31]:
# IIP annual, 2023-24
nsoindia.get_data("IIP", {
    "base_year": "2011-12",
    "financial_year": "2023-24",
}, format="df")

,base_year,year,type,category,sub_category,index,growth_rate
0,2011-12,2023-24,General,General,,146.7,5.9
1,2011-12,2023-24,Sectoral,Mining,,128.9,7.5
2,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Food Products,134.5,1.6
3,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Beverages,110.9,5.1
4,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Tobacco Products,81.1,-8.3
5,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Textiles,107.6,0.1
6,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Wearing Apparel,109.9,-14.1
7,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Leather and Related Products,95.0,-1.0
8,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Wood and Products of Wood and C...,98.3,-5.9
9,2011-12,2023-24,Sectoral,Manufacturing,Manufacture of Paper and Paper Products,79.4,-3.6


---
## WPI - Wholesale Price Index

Hierarchical commodity structure (1000+ items). No indicator code needed - `get_metadata` returns the full commodity hierarchy.

In [32]:
# WPI commodity hierarchy
metadata = nsoindia.get_metadata("WPI")
metadata

{'year': [{'year': 2012},
  {'year': 2013},
  {'year': 2014},
  {'year': 2015},
  {'year': 2016},
  {'year': 2017},
  {'year': 2018},
  {'year': 2019},
  {'year': 2020},
  {'year': 2021},
  {'year': 2022},
  {'year': 2023},
  {'year': 2024},
  {'year': 2025},
  {'year': 2026}],
 'month': [{'month_code': 1, 'month_name': 'January'},
  {'month_code': 2, 'month_name': 'February'},
  {'month_code': 3, 'month_name': 'March'},
  {'month_code': 4, 'month_name': 'April'},
  {'month_code': 5, 'month_name': 'May'},
  {'month_code': 6, 'month_name': 'June'},
  {'month_code': 7, 'month_name': 'July'},
  {'month_code': 8, 'month_name': 'August'},
  {'month_code': 9, 'month_name': 'September'},
  {'month_code': 10, 'month_name': 'October'},
  {'month_code': 11, 'month_name': 'November'},
  {'month_code': 12, 'month_name': 'December'}],
 'major_group': [{'major_group_code': '1100000000',
   'major_group_name': 'Primary articles',
   'viz': 'line'},
  {'major_group_code': '1200000000',
   'major_group

In [33]:
# WPI for December 2024
nsoindia.get_data("WPI", {
    "year": "2024",
    "month_code": "12",
}, format="df")

,year,month,majorgroup,group,subgroup,sub_subgroup,item,index_value
0,2024,December,Wholesale Price Index,None,None,None,None,155.7
1,2024,December,Primary articles,None,None,None,None,193.8
2,2024,December,Primary articles,Food articles,None,None,None,207.5
3,2024,December,Primary articles,Food articles,Food grains (cereals+pulses),None,None,213.7
4,2024,December,Primary articles,Food articles,Food grains (cereals+pulses),Cereals,None,211.4
5,2024,December,Primary articles,Food articles,Food grains (cereals+pulses),Cereals,Paddy,205.3
6,2024,December,Primary articles,Food articles,Food grains (cereals+pulses),Cereals,Wheat,215.5
7,2024,December,Primary articles,Food articles,Food grains (cereals+pulses),Cereals,Jowar,201.4
8,2024,December,Primary articles,Food articles,Food grains (cereals+pulses),Cereals,Bajra,255.1
9,2024,December,Primary articles,Food articles,Food grains (cereals+pulses),Cereals,Maize,212.9


---
## GENDER - Gender Statistics

147 indicators covering demographics, health, education, labour, financial inclusion, and crimes against women.

In [34]:
indicators = nsoindia.get_indicators("GENDER")
print(f"Total indicators: {len(indicators)}\n")
for i in indicators[:10]:
    print(f"  {i['indicator_code']}: {i['description']}")

Total indicators: 147

  1: Trend in Population
  2: Sex Ratio
  3: Average Annual Growth Rate of Population
  4: Percentage Distribution of Projected Population
  5: Sex ratio in context of developmental planning
  6: State-wise Sex Ratio 
  7: Sex ratio at birth for bigger States based on Sample Registration System
  8: General Fertility Rate
  9: Age-specific Fertility rate
  10: Total Fertility Rate 


In [ ]:
# Sex ratio (indicator_code=2)
nsoindia.get_data("GENDER", {"indicator_code": 2}, format="df")

---
## ENVSTATS - Environment Statistics

124 indicators covering climate, water, forests, biodiversity, pollution, and natural disasters.

In [ ]:
indicators = nsoindia.get_indicators("ENVSTATS")
print(f"Total indicators: {len(indicators)}\n")
for i in indicators[:10]:
    print(f"  {i['indicator_code']}: {i.get('description', i.get('label', ''))}")

Total indicators: 124

  1: Annual and Seasonal Mean Temperature
  2: Annual and Seasonal Minimum and Maximum Temperature
  3: Annual and Monthly Rainfall
  4: Wetland distribution in India
  5: Number and Area of Wetlands
  6: Water Sheds in India
  7: Catchment Area of major River Basins
  8: Length of Coastline and Coastal Population of India
  9: Occurrence of Major Soils in India
  10: Category-wise distribution of land degradation in different States of India


In [ ]:
# Mean temperature trend (indicator_code=1)
nsoindia.get_data("ENVSTATS", {"indicator_code": 1}, format="df")

---
## RBI - Foreign Trade and Exchange Rates

39 indicators on balance of payments, forex reserves, exchange rates (155 currencies), NRI deposits.

In [ ]:
indicators = nsoindia.get_indicators("RBI")
for i in indicators:
    print(f"  {i['indicator_code']}: {i['label']}")

  22: Balance of Payments ‑ Indicators
  24: Broad Commodity Composition of India's Merchandise Trade ‑ Oil and Non oil Exports and Imports ‑ Rupees
  42: Broad Commodity Composition of India’s Merchandise Trade - Oil and Non-Oil Exports and Imports - US Dollar
  1: Direction of Foreign Trade ‑ Rupees
  11: Direction of Foreign Trade ‑ US Dollars
  32: Exchange Rate of the Indian Rupee Vis‑a‑Vis the SDR, US Dollar, Pound Sterling, D.M./Euro and Japanese Yen (Calendar Year ‑ Annual Average)
  31: Exchange Rate of the Indian Rupee Vis‑A‑Vis the SDR, US Dollar, Pound Sterling, D.M./Euro and Japanese Yen (Financial Year ‑ Annual Average and End‑Year Rates)
  35: Exchange Rate of the Indian Rupee vis-à-vis the SDR, US Dollar, Pound Sterling, Euro and Japanese Yen (High and Low During the Month)
  33: Exchange Rate of the Indian Rupee vis-à-vis the SDR, US Dollar, Pound Sterling, Euro and Japanese Yen (Monthly Average and end-Month Rates)
  43: EXPORTS OF MAJOR COMMODITIES - RUPEES
  44: EXP

In [ ]:
# Get valid sub_indicator_codes for Balance of Payments (indicator_code=22)
metadata = nsoindia.get_metadata("RBI", sub_indicator_code=22)
metadata

In [ ]:
# Balance of Payments
nsoindia.get_data("RBI", {"indicator_code": 22}, format="df")

---
## EC - Economic Census

District-level establishment and worker counts across three census editions (EC4/EC5/EC6).

In [ ]:
nsoindia.get_indicators("EC", format="df")

In [ ]:
# EC6 (2013-14) - available states and filter options
metadata = nsoindia.get_metadata("EC", indicator_code=1)
metadata

In [ ]:
# EC6 - top districts in Maharashtra by establishments
nsoindia.get_data("EC", {
    "indicator_code": 1,
    "state": "27",       # Maharashtra
    "mode": "ranking",
    "top5opt": "2",
}, format="df")

---
## Output Formats

All functions support `format="dict"` (default), `format="df"` (pandas DataFrame), or `format="csv"`.

In [ ]:
csv_str = nsoindia.get_data("PLFS", {
    "indicator_code": 3,
    "frequency_code": 1,
    "year": "2023-24",
    "state_code": 99,
    "gender_code": 3,
    "age_code": 1,
    "sector_code": 3,
}, format="csv")

print(csv_str[:600])

year,frequency,indicator,state,gender,sector,AgeGroup,weekly_status,religion,socialGroup,General_Education,quarter,month,value,unit
2023-24,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,all,all,,3.2,%
2023-24,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,CWS,all,all,all,all,,4.9,%
2023-24,Annually,"UR (Unemployment Rate, in per cent)",All India,person,rural + urban,15 years and above,PS+SS,all,all,1.not literate,all,,0.2,%
2023-24,Annually,"UR (Unemployment Rate, in per cent)


---
## Error Handling

In [ ]:
# Typo in dataset name - fuzzy suggestion
try:
    nsoindia.get_indicators("plfs_data")
except InvalidDatasetError as e:
    print(e)

Unknown dataset: 'plfs_data'. Valid datasets: PLFS, CPI, IIP, ASI, NAS, WPI, ENERGY, AISHE, ASUSE, GENDER, NFHS, ENVSTATS, RBI, NSS77, NSS78, CPIALRL, HCES, TUS, EC. Did you mean: PLFS, ENVSTATS, NFHS?


In [ ]:
# Missing required parameter
try:
    nsoindia.get_metadata("PLFS", frequency_code=1)
except InvalidFilterError as e:
    print(e)

Missing required params for get_metadata(dataset='PLFS'): ['indicator_code']


In [ ]:
# Invalid filter parameter name
try:
    nsoindia.get_data("PLFS", {
        "indicator_code": 1,
        "frequency_code": 1,
        "typo_param": "hello",
    })
except InvalidFilterError as e:
    print(e)

Invalid params: ['typo_param']. Valid params: ['indicator_code', 'frequency_code', 'year', 'month_code', 'state_code', 'gender_code', 'age_code', 'sector_code', 'quarter_code', 'weekly_status_code', 'religion_code', 'social_category_code', 'education_code', 'broad_industry_work_code', 'broad_status_employment_code', 'employee_contract_code', 'enterprise_size_code', 'enterprise_type_code', 'industry_section_code', 'nco_division_code', 'nic_group_code', 'limit', 'page', 'Format'].
